File: packages/typescript/goldenpipe/src/core/engine/runner.ts
/**
 * Pipeline runner — execute stages with error handling and routing.
 * Port of goldenpipe/engine/runner.py.
 *
 * ASYNC: stage execution awaits each stage's `run`, because the GoldenMatch
 * `dedupe` adapter is async.
 *
 * Edge-safe: no `node:` imports.
 */

import type { PipeContext, StageResult } from "../models.js";
import { StageStatus } from "../models.js";
import type { StageRegistry } from "./registry.js";
import type { ExecutionPlan } from "./resolver.js";
import { Router } from "./router.js";

function isFalsy(value: unknown): boolean {
  if (value === null || value === undefined) return true;
  if (value === false || value === 0 || value === "") return true;
  if (Array.isArray(value)) return value.length === 0;
  if (value instanceof Map || value instanceof Set) return value.size === 0;
  if (typeof value === "object") return Object.keys(value).length === 0;
  return false;
}

export class Runner {
  constructor(private readonly registry: StageRegistry) {}

  /** Execute an ExecutionPlan against a PipeContext, returning per-stage results. */
  async run(plan: ExecutionPlan, ctx: PipeContext): Promise<Record<string, StageResult>> {
    const results: Record<string, StageResult> = {};
    let remaining = [...plan.stages];

    while (remaining.length > 0) {
      const planned = remaining.shift()!;

      if (planned.spec.skipIf) {
        const artifact = ctx.artifacts[planned.spec.skipIf];
        if (isFalsy(artifact)) {
          results[planned.name] = { status: StageStatus.SKIPPED };
          ctx.reasoning[planned.name] =
            `Skipped: artifact '${planned.spec.skipIf}' is missing/falsy`;
          continue;
        }
      }

      const start = performance.now();
      try {
        // Make stage-level config available to the adapter via context.
        ctx.stageConfig = planned.config;

        await planned.stage.validate(ctx);
        const result = await planned.stage.run(ctx);
        ctx.timing[planned.name] = (performance.now() - start) / 1000;
        results[planned.name] = result;

        if (result.decision != null) {
          remaining = Router.apply(result.decision, remaining, ctx, this.registry);
        }
      } catch (e) {
        ctx.timing[planned.name] = (performance.now() - start) / 1000;
        const message = e instanceof Error ? e.message : String(e);
        results[planned.name] = { status: StageStatus.FAILED, error: message };
        ctx.reasoning[planned.name] = `Failed: ${message}`;

        if (planned.spec.onError === "abort") {
          break;
        }
      }
    }

    return results;
  }
}

File: packages/python/goldenpipe/goldenpipe/engine/runner.py
"""Pipeline runner -- execute stages with error handling and routing."""
from __future__ import annotations

import logging
import time

from goldenpipe.engine.registry import StageRegistry
from goldenpipe.engine.resolver import ExecutionPlan
from goldenpipe.engine.router import Router
from goldenpipe.models.context import PipeContext, StageResult, StageStatus

logger = logging.getLogger(__name__)


class Runner:
    """Executes an ExecutionPlan against a PipeContext."""

    def __init__(self, registry: StageRegistry) -> None:
        self._registry = registry

    def run(self, plan: ExecutionPlan, ctx: PipeContext) -> dict[str, StageResult]:
        results: dict[str, StageResult] = {}
        remaining = list(plan.stages)

        while remaining:
            planned = remaining.pop(0)

            if planned.spec.skip_if:
                artifact = ctx.artifacts.get(planned.spec.skip_if)
                if not artifact:
                    result = StageResult(status=StageStatus.SKIPPED)
                    results[planned.name] = result
                    ctx.reasoning[planned.name] = (
                        f"Skipped: artifact '{planned.spec.skip_if}' is missing/falsy"
                    )
                    continue

            # Relocatable-stage seam (contract Phases A/C).
            location = getattr(getattr(planned.stage, "info", None), "location", "local")
            if location == "local":
                # Transition back to Python: if the previous (remote) stage left an
                # engine-resident frame, materialize it here -- the boundary
                # crossing, paid once, exactly when a local stage needs the data.
                engine = getattr(ctx, "_frame", None)
                if engine is not None:
                    ctx.df = engine.polars()
                    ctx._frame = None
            elif not getattr(planned.stage, "remote_capable", False):
                # A stage declares a remote location but provides no remote
                # implementation -- a not-yet-built placement. Fail loudly (before
                # the try/except) rather than silently running it in-process. A
                # real RemoteStage sets ``remote_capable = True`` and runs below.
                raise NotImplementedError(
                    f"Stage '{planned.name}' declares location={location!r} but is not a "
                    "RemoteStage (remote_capable). Remote execution for it is not "
                    "implemented (relocatable-stage contract Phase C)."
                )

            start = time.perf_counter()
            try:
                # Make stage-level config available to the adapter via context
                ctx.stage_config = planned.config

                if hasattr(planned.stage, "validate") and callable(planned.stage.validate):
                    planned.stage.validate(ctx)
                result = planned.stage.run(ctx)
                elapsed = time.perf_counter() - start
                ctx.timing[planned.name] = elapsed
                results[planned.name] = result

                if result.decision is not None:
                    remaining = Router.apply(
                        result.decision, remaining, ctx, self._registry,
                    )

            except Exception as e:
                import traceback as _tb
                logger.error("Stage %s failed:\n%s", planned.name, _tb.format_exc())
                elapsed = time.perf_counter() - start
                ctx.timing[planned.name] = elapsed
                result = StageResult(status=StageStatus.FAILED, error=str(e))
                results[planned.name] = result
                ctx.reasoning[planned.name] = f"Failed: {e}"

                if planned.spec.on_error == "abort":
                    break

        return results
